### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by JasonZhongZexin**

### CODE REVIEW

File: include/SVF-FE/LLVMUtil.h [hunk 1]
```diff
@@ -160,12 +160,14 @@ inline bool ArgInNoCallerFunction (const Value * val)
...
 /// Get reachable basic block from function entry
-void getFunReachableBBs (const Function * fun, DominatorTree* dt,std::vector<const BasicBlock*>& bbs);
+void getFunReachableBBs (const SVFFunction* svffun, std::vector<const BasicBlock*>& bbs);
 
 /// Get function exit basic block
 /// FIXME: this back() here is only valid when UnifyFunctionExitNodes pass is invoked
 inline const BasicBlock* getFunExitBB(const Function* fun)
 {
+    const SVFFunction* svffun = LLVMModuleSet::getLLVMModuleSet()->getSVFFunction(fun);
+    assert(SVFUtil::isExtCall(svffun) == false);
```
Comment (mbarbar) [inline include/SVF-FE/LLVMUtil.h:170]:
```
!SVFUtil::isExtCall
```
not
```
== false
```

File: include/SVF-FE/LLVMUtil.h [hunk 2]
```diff
@@ -160,12 +160,14 @@ inline bool ArgInNoCallerFunction (const Value * val)
 bool functionDoesNotRet (const Function * fun);
 
 /// Get reachable basic block from function entry
-void getFunReachableBBs (const Function * fun, DominatorTree* dt,std::vector<const BasicBlock*>& bbs);
+void getFunReachableBBs (const SVFFunction* svffun, std::vector<const BasicBlock*>& bbs);
```
Comment (mbarbar) [inline include/SVF-FE/LLVMUtil.h:163]: `fun` or `svfFun` is better.

File: include/Util/SVFModule.h [hunk 3]
```diff
@@ -65,6 +65,19 @@ class SVFModule
...
+    Set<const Function*> isDeadFunction;
+    Set<const Instruction*> isReturns;
+    Set<const Value*> nullPtrSyms;
+    Set<const Value*> blackholeSyms;
+    Set<const Function*> functionDoesNotRet;
+    Set<const Value*> isPtrInDeadFunction;
+    Map<const Function*, const BasicBlock*> funExitBBMap;
+    Map<const BasicBlock*, const u32_t> BBSuccessorNumMap;
+    Map<const PointerType*, const Type*> ptrElementTypeMap;
+    Map<const BasicBlock*, const Map<const BasicBlock*, const u32_t>> BBSuccessorPosMap;
```
Comment (mbarbar) [inline include/Util/SVFModule.h:78]: Rename to `bbSuccessorPosMap`, same with others (`BB` -> `bb`)

File: include/Util/SVFModule.h [hunk 4]
```diff
@@ -65,6 +65,19 @@ class SVFModule
...
+    Map<const BasicBlock*, const Map<const BasicBlock*, const u32_t>> BBPredecessorPosMap;
```
Comment (mbarbar) [inline include/Util/SVFModule.h:79]: What is this mapping from/to?

File: include/Util/SVFModule.h [hunk 5]
```diff
@@ -209,6 +221,142 @@ class SVFModule
...
+
+    inline const Map<const PointerType*, const Type*> getPtrElementTypeMap()
+    {
+        return ptrElementTypeMap;
+    }
+
+    inline void addFunExitBB(const Function *fun, const BasicBlock* bb)
+    {
+        const SVFFunction* svffun = getSVFFunction(fun);
+        assert((fun && ExtAPI::getExtAPI()->is_ext(svffun)) == false);
```
Comment (mbarbar) [inline include/Util/SVFModule.h:286]: Use `!`, not `== false`

File: include/Util/SVFModule.h [hunk 6]
```diff
@@ -209,6 +221,142 @@ class SVFModule
...
+    }
+
+    inline const Map<const PointerType*, const Type*> getPtrElementTypeMap()
+    {
+        return ptrElementTypeMap;
+    }
+
+    inline void addFunExitBB(const Function *fun, const BasicBlock* bb)
+    {
+        const SVFFunction* svffun = getSVFFunction(fun);
```
Comment (mbarbar) [inline include/Util/SVFModule.h:285]: `svfFun` is better

File: include/Util/SVFModule.h [hunk 7]
```diff
@@ -209,6 +221,142 @@ class SVFModule
...
+             BBSuccessorPosMap.insert({bb,foundValue});
+        } 
+        else {
+            Map<const BasicBlock*, const u32_t> valueMap;
+            valueMap.insert({succ,pos});
+            BBSuccessorPosMap.insert({bb,valueMap});
+        }
+    }
+
+    inline void addBBPredecessorPos(const BasicBlock *bb, const BasicBlock* Pred,const u32_t pos)
```
Comment (mbarbar) [inline include/Util/SVFModule.h:309]: `pred` not `Pred`

File: include/Util/SVFModule.h [hunk 8]
```diff
@@ -209,6 +221,142 @@ class SVFModule
...
+
+    inline void addBBSuccessorNum(const BasicBlock *bb, const u32_t num)
+    {
+        BBSuccessorNumMap.insert({bb,num});
+    }
+
+    inline void addBBSuccessorPos(const BasicBlock *bb, const BasicBlock* succ,const u32_t pos)
+    {
+        if(BBSuccessorPosMap.find(bb) != BBSuccessorPosMap.end()){
+            Map<const BasicBlock*, const u32_t> foundValue = BBSuccessorPosMap.find(bb)->second;
```
Comment (mbarbar) [inline include/Util/SVFModule.h:298]: This should be cached from the initial lookup -- double is unnecessary.

Comment (mbarbar) [reply include/Util/SVFModule.h:298]: Same as pred

File: lib/MemoryModel/SymbolTableInfo.cpp [hunk 9]
```diff
@@ -323,15 +321,93 @@ void SymbolTableInfo::destroy()
...
+bool SymbolTableInfo::isPtrInDeadFunction (const Value * value)
+{
+    return symInfo->getModule()->getPtrInDeadFunction().find(value) != symInfo->getModule()->getPtrInDeadFunction().end();
+}
+
+const BasicBlock* SymbolTableInfo::getFunExitBB(const Function* fun)
+{
+    if (symInfo->getModule()->getFunExitBBMap().find(fun) != symInfo->getModule()->getFunExitBBMap().end())
+    {
+        return symInfo->getModule()->getFunExitBBMap().find(fun)->second;
```
Comment (mbarbar) [inline lib/MemoryModel/SymbolTableInfo.cpp:356]: Double lookup unnecessary.

File: lib/MemoryModel/SymbolTableInfo.cpp [hunk 10]
```diff
@@ -323,15 +321,93 @@ void SymbolTableInfo::destroy()
...
+    }
+    return NULL;
+}
+
+const u32_t SymbolTableInfo::getBBSuccessorNum(const BasicBlock *bb)
+{
+    if (symInfo->getModule()->getBBSuccessorNumMap().find(bb) != symInfo->getModule()->getBBSuccessorNumMap().end())
+    {
+        return symInfo->getModule()->getBBSuccessorNumMap().find(bb)->second;
```
Comment (mbarbar) [inline lib/MemoryModel/SymbolTableInfo.cpp:365]: Double lookup unnecessary.

File: lib/MemoryModel/SymbolTableInfo.cpp [hunk 11]
```diff
@@ -323,15 +321,93 @@ void SymbolTableInfo::destroy()
...
+    }
+    return 0;
+}
+
+const u32_t SymbolTableInfo::getBBSuccessorPos(const BasicBlock *bb, const BasicBlock *succ)
+{
+    if(symInfo->getModule()->getBBSuccessorPosMap().find(bb) != symInfo->getModule()->getBBSuccessorPosMap().end()){
+        const Map <const BasicBlock *, const u32_t> value = symInfo->getModule()->getBBSuccessorPosMap().find(bb)->second;
```
Comment (mbarbar) [inline lib/MemoryModel/SymbolTableInfo.cpp:373]: Double lookup unnecessary

File: lib/MemoryModel/SymbolTableInfo.cpp [hunk 12]
```diff
@@ -323,15 +321,93 @@ void SymbolTableInfo::destroy()
...
+            return pos;
+        }
+    }
+    return 0;
+}
+
+const u32_t SymbolTableInfo::getBBPredecessorPos(const BasicBlock *bb, const BasicBlock *Pred)
+{
+    if(symInfo->getModule()->getBBPredecessorPosMap().find(bb) != symInfo->getModule()->getBBPredecessorPosMap().end()){
+        const Map <const BasicBlock *, const u32_t> value = symInfo->getModule()->getBBPredecessorPosMap().find(bb)->second;
```
Comment (mbarbar) [inline lib/MemoryModel/SymbolTableInfo.cpp:386]: Double lookup unnecessary

File: lib/MemoryModel/SymbolTableInfo.cpp [hunk 13]
```diff
@@ -323,15 +321,93 @@ void SymbolTableInfo::destroy()
...
+        if(value.find(succ) != value.end())
+        {
+            u32_t pos = value.find(succ)->second;
+            return pos;
+        }
+    }
+    return 0;
+}
+
+const u32_t SymbolTableInfo::getBBPredecessorPos(const BasicBlock *bb, const BasicBlock *Pred)
```
Comment (mbarbar) [inline lib/MemoryModel/SymbolTableInfo.cpp:383]: `pred` not `Pred`

File: lib/MemoryModel/SymbolTableInfo.cpp [hunk 14]
```diff
@@ -323,15 +321,93 @@ void SymbolTableInfo::destroy()
...
+        }
+    }
+    return 0;
+}
+
+const Type* SymbolTableInfo::getPtrElementType(const PointerType* pty)
+{
+    if (symInfo->getModule()->getPtrElementTypeMap().find(pty) != symInfo->getModule()->getPtrElementTypeMap().end())
+    {
+        return symInfo->getModule()->getPtrElementTypeMap().find(pty)->second;
```
Comment (mbarbar) [inline lib/MemoryModel/SymbolTableInfo.cpp:400]: Double lookup unnecessary.

File: lib/SVF-FE/LLVMUtil.cpp [hunk 15]
```diff
@@ -91,17 +91,22 @@ bool LLVMUtil::isConstantObjSym(const Value *val)
 /*!
  * Return reachable bbs from function entry
  */
-void LLVMUtil::getFunReachableBBs (const Function * fun, DominatorTree* dt, std::vector<const BasicBlock*> &reachableBBs)
+void LLVMUtil::getFunReachableBBs (const SVFFunction* svffun, std::vector<const BasicBlock*> &reachableBBs)
```
Comment (mbarbar) [inline lib/SVF-FE/LLVMUtil.cpp:94]: `svfFun`

File: lib/SVF-FE/LLVMUtil.cpp [hunk 16]
```diff
@@ -91,17 +91,22 @@ bool LLVMUtil::isConstantObjSym(const Value *val)
...
 {
+    assert(SVFUtil::isExtCall(svffun) == false);
```
Comment (mbarbar) [inline lib/SVF-FE/LLVMUtil.cpp:96]: `!` vs. `== false`

File: lib/SVF-FE/SymbolTableBuilder.cpp [hunk 17]
```diff
@@ -205,9 +205,80 @@ void SymbolTableBuilder::buildMemModel(SVFModule* svfModule)
...
     if (LLVMUtil::isBlackholeSym(val))
-        symInfo->blackholeSyms.insert(val);
+        symInfo->getModule()->addBlackholeSyms(val);
+}
+
+
+void SymbolTableBuilder::collectSpecialSym(const Value* val){
+    if (const Function *fun = SVFUtil::dyn_cast<Function>(val))
+    {
+        const SVFFunction *svffun = symInfo->getModule()->getSVFFunction(fun);
```
Comment (mbarbar) [inline lib/SVF-FE/SymbolTableBuilder.cpp:217]: `svfFun`

File: lib/SVF-FE/SymbolTableBuilder.cpp [hunk 18]
```diff
@@ -205,9 +205,80 @@ void SymbolTableBuilder::buildMemModel(SVFModule* svfModule)
...
+            symInfo->getModule()->addFunExitBB(fun,exitBB);
+
+            for (Function::const_iterator bit = fun->begin(), ebit = fun->end(); bit != ebit; ++bit)
+            {
+                const BasicBlock *bb = &*bit;
+                const u32_t num = LLVMUtil::getBBSuccessorNum(bb);
+                symInfo->getModule()->addBBSuccessorNum(bb,num);
+                if (num >1)
+                {
+                    for (succ_const_iterator succ_it = succ_begin(bb); succ_it != succ_end(bb); succ_it++)
```
Comment (mbarbar) [inline lib/SVF-FE/SymbolTableBuilder.cpp:238]: `succIt`

File: lib/SVF-FE/SymbolTableBuilder.cpp [hunk 19]
```diff
@@ -205,9 +205,80 @@ void SymbolTableBuilder::buildMemModel(SVFModule* svfModule)
...
+                    {
+                        const BasicBlock* succ = *succ_it;
+                        const u32_t successorPos = LLVMUtil::getBBSuccessorPos(bb,succ);
+                        if (successorPos != 0)
+                        {
+                            symInfo->getModule()->addBBSuccessorPos(bb,succ,successorPos);
+                        }
+                    }
+                }
+                for (succ_const_iterator succ_it = succ_begin(bb); succ_it != succ_end(bb); succ_it++)
```
Comment (mbarbar) [inline lib/SVF-FE/SymbolTableBuilder.cpp:248]: `succIt`

File: lib/SVF-FE/SymbolTableBuilder.cpp [hunk 20]
```diff
@@ -222,10 +293,12 @@ void SymbolTableBuilder::collectSym(const Value *val)
 
     //TODO handle constant expression value here??
     handleCE(val);
-
+
```
Comment (mbarbar) [inline lib/SVF-FE/SymbolTableBuilder.cpp:296]: no need

File: include/Util/SVFModule.h [hunk 21]
```diff
@@ -101,9 +114,8 @@ class SVFModule
     void buildSymbolTableInfo();
 
     ///@{
-    inline void addFunctionSet(Function* fun)
+    inline void addFunctionSet(Function* fun, SVFFunction* svfFunc)
```
Comment (yuleisui) [inline include/Util/SVFModule.h:117]: remove Function as much as possible

File: include/SVF-FE/LLVMUtil.h (hunk 1 above)
Comment (yuleisui) [inline include/SVF-FE/LLVMUtil.h:170]: put a message `&&`

File: lib/SVF-FE/SymbolTableBuilder.cpp [hunk 22]
```diff
@@ -273,11 +277,6 @@ void SymbolTableBuilder::collectSpecialSym(const Value* val){
        {
```
Comment (yuleisui) [inline lib/SVF-FE/SymbolTableBuilder.cpp:277]: else if (LLVMUtil::ArgInNoCallerFunction(val)){
...
}

**Pull request closed without merging**